ISSUE #590 (@bug W-4556980@) Reduce excessive CPU usage on client side: add crc3…

dlg99 opened the pull request:
…2c with SSE processor instruction support, benchmark to make sure it is better than crc32

Descriptions of the changes in this PR:

CRC32C for bookkeeper

Master Issue: #590

merlimat commented on microbenchmarks/pom.xml:
```diff
@@ -0,0 +1,151 @@
+<!-- Licensed to the Apache Software Foundation (ASF) under one or more contributor 
```
AFS header appears to be misformatted.

dlg99 replied:
will fix

merlimat commented on microbenchmarks/results.txt:
```diff
@@ -0,0 +1,89 @@
+# on macbook pro
```
@dlg99 One thing to note is that the SSE4.2 instructions are using a JNI module and that's only compiled for linux in the released artifacts for `pulsar-checksum`.

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerMetadata.java:
```diff
@@ -119,7 +119,7 @@ public LedgerMetadata(int ensembleSize,
 
         this.digestType = digestType.equals(BookKeeper.DigestType.MAC)
-            ? LedgerMetadataFormat.DigestType.HMAC : LedgerMetadataFormat.DigestType.CRC32;
+            ? LedgerMetadataFormat.DigestType.HMAC : LedgerMetadataFormat.DigestType.valueOf(digestType.toString());
```
Can this be simplified with `this.digestType = LedgerMetadataFormat.DigestType.valueOf(digestType.toString())`?

dlg99 replied:
@sijie I wish, there is "HMAC" vs "MAC" case.

sijie replied:
gotcha

sijie commented on microbenchmarks/pom.xml:
```diff
@@ -0,0 +1,151 @@
+        <artifactId>bookkeeper</artifactId>
+        <version>4.7.0-SNAPSHOT</version>
+    </parent>
+    <artifactId>microbenchmarks</artifactId>
```
There is actually a module `jmh` under `tests`: https://github.com/apache/bookkeeper/tree/master/tests/jmh

We can push this jmh benchmark in that module.

dlg99 replied:
I added this one for a few reasons:
- I already had it like that ;)
- jmh-bench did not build even after I fixed version in pom. "ProtocolBenchmark.java:[88,41] constructor AddRequest in class org.apache.bookkeeper.proto.BookieProtocol.AddRequest cannot be applied to given types". 
- I was not sure what the plans fro jmh-bench are, given its current state.

sijie replied:
don't worry at this patch. I think jmh module was forgot due to some merge issues before. I had the change to fix that. I can incorporate your module in my change.

sijie commented on microbenchmarks/results.txt:
```diff
@@ -0,0 +1,89 @@
+# on macbook pro
```
we probably don't need this file to be checked in.

dlg99 replied:
@sijie It takes awhile to get results, I assume we don't want to run these on each build.
OTOH having benchmark results easily accessible may help people decide on digest type to use. 
I can remove it, of course, if you insist.

sijie replied:
@dlg99 okay for me.

merlimat replied:
I think it's good to have it somewhere. To help people decide, probably we should also mention a summary of the differences in the Javadoc and website.

sijie commented on microbenchmarks/run.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
```
add license header?

dlg99 replied:
will do

sijie commented on bookkeeper-server/pom.xml:
```diff
@@ -182,6 +182,11 @@
       <artifactId>http-server</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
```
nit: I think we probably want to move the whole `checksum` package from pulsar to bookkeeper. but it is okay for this pull request. I will discuss with @merlimat on how to handle that.

dlg99 replied:
@sijie I'll address the comments later/tomorrow-ish.

On the other note I also tried/compared adler32 (from jdk) and crc32c from jdk9 (ported it to run with jdk8), neither were better than regular crc32.
oh, and I tried DUMMY digest (aka no digest). This one crazy fast and garbage free. Unfortunately, it is useless.

sijie replied:
@dlg99 did you run with jdk9? I've seem very good number with jdk9. in java 9, the crc32c with sse42 is intrinsic on jvm. I will send my results as well.

merlimat replied:
In the checksum module we should be able to check for the JVM runtime version and select the fastest available implementation.

dlg99 replied:
@sijie I ran with jdk8. I tweaked http://hg.openjdk.java.net/jdk9/dev/jdk/file/feef46bbf70b/src/java.base/share/classes/java/util/zip/CRC32C.java to work on jdk8 and, to my surprise, saw no performance difference compared with jdk8's crc32.

sijie replied:
@dlg99 the approach you took didn't work because the assemble code of using intel crc32c instruction is a hotspot code which only available in jdk9. you can port the code from java9 to java8 and it can compile and work, but jdk8 doesn't this assemble code, so why you won't see any performance difference.

You can checkout the hotspot code here http://hg.openjdk.java.net/jdk9/jdk9/hotspot/rev/a60e232aa8f2#l11.20

sijie replied:
Also check my benchmark at #858

jvrao replied:
crc32c will have its fullest potential only if it is using hardware instruction set. (obviously) :)

sijie replied:
Created #915 for tracking importing the pulsar-checksum module to bookkeeper as discussed with @merlimat before. 

It *doesn't* block this PR. This PR can go in first if we agreed with 4 bytes code length with crc32c.

merlimat commented on microbenchmarks/src/main/java/org/apache/bookkeeper/client/DigestTypeBenchmark.java:
```diff
@@ -0,0 +1,172 @@
+
+            digestBuf = Unpooled.buffer(getDigestManager(digest).getMacCodeLength());
+
+            final int entrySizeKb = 64;
```
Might be worth comparing also with different entry sizes.

dlg99 replied:
will do

merlimat commented on microbenchmarks/results.txt:
```diff
@@ -0,0 +1,89 @@
+DigestTypeBenchmark.digestManager:·stack                                  NOT_ARRAY_BACKED     CRC32  thrpt             NaN                ---
+DigestTypeBenchmark.digestManager                                         NOT_ARRAY_BACKED   CRC32_C  thrpt    5    130.444 ±    3.441  ops/ms
+DigestTypeBenchmark.digestManager:·gc.alloc.rate                          NOT_ARRAY_BACKED   CRC32_C  thrpt    5   7824.727 ±  212.028  MB/sec
+DigestTypeBenchmark.digestManager:·gc.alloc.rate.norm                     NOT_ARRAY_BACKED   CRC32_C  thrpt    5  65552.001 ±    0.001    B/op
```
Having 64Kb allocated (and GCed) per each buffer seems to indicate that something is off. Prob, that's because of the pure java implementation fallback due to not having the JNI lib on MacOS.

dlg99 replied:
@merlimat in unit tests I did not see this log: 
        if (!Sse42Crc32C.isSupported()) {
            LOG.error("Sse42Crc32C is not supported, will use less slower CRC32C implementation.");
        }
so I assume that it won't use pure java implementation.

This is because DoubleByteBuf returns  false == hasMemoryAddress() and false == hasArray()

    public static int computeChecksum(ByteBuf payload) {
        if (payload.hasMemoryAddress() && (CRC32C_HASH instanceof Sse42Crc32C)) {
            return CRC32C_HASH.calculate(payload.memoryAddress() + payload.readerIndex(), payload.readableBytes());
        } else if (payload.hasArray()) {
            return CRC32C_HASH.calculate(payload.array(), payload.arrayOffset() + payload.readerIndex(),
                    payload.readableBytes());
        } else {
            return CRC32C_HASH.calculate(payload.nioBuffer());
        }
    }

ends up in .nioBuffer() which ends up doing the following:

    public ByteBuffer nioBuffer(int index, int length) {
      >>>>>   ByteBuffer dst = ByteBuffer.allocate(length);
        ByteBuf b = Unpooled.wrappedBuffer(dst);

merlimat replied:
Oh, that's true. That's because it's doing that on a `DoubleByteBuf`. We should make sure we don't do that in the real code :)

dlg99 commented:
@sijie @merlimat I've addressed the CR feedback. 
Please let me know if you want me to squash commits together, I assumed this will be done as part of pull request's merge.

dlg99 commented:
Build failure is unrelated to my changes:

"[ERROR] /home/jenkins/jenkins-slave/workspace/bookkeeper-precommit-pullrequest/bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java:[73,8] org.apache.bookkeeper.bookie.storage.ldb.DbLedgerStorage is not abstract and does not override abstract method "

sijie commented:
@dlg99 that was due two concurrent merges. the first one change the method, the second one was introducing a new implmenetation based on the old interface. it should be fixed now. I am triggering ci builds.

sijie commented:
retest this please

sijie commented:
close/reopen the pull request to trigger travis ci build as well.

dlg99 commented:
@sijie Jenkins succeeded. 
Travis CI was runs rat plugin which was not happy to see results.txt without apache header.
I added header and squashed commits. Hopefully this takes care of things.

What is the right way to run maven locally to hit all these checks? "mvn clean install" is not enough to mimic travis's behavior.

sijie commented:
cool.
```
maven clean apache-rat:check install spotbugs:check
```

`apache-rat:check` checks the license headers
`spotbugs:check` checks the bugs

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/client/CRC32CDigestManager.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Override
+    int getMacCodeLength() {
+        return 4;
```
@merlimat @dlg99 shall we stick to 8 like crc32 for consistency? assume we might use jdk9 crc32c in future.

https://docs.oracle.com/javase/9/docs/api/java/util/zip/CRC32C.html#getValue--

dlg99 replied:
@sijie Is there any guarantee that these will be interchangeable? i.e. jdk's one may  use actual full long and pulsar's one will have int + another int (i.e. 0) as a filler. So we won't be able to upgrade bookie if it has ledgers created with one crc32c implementation and bookie just switched to another. Safer way would be to have yet another digest type (CRC32C_J) and set it explicitly.

sijie replied:
@dlg99 I think it should be exchangeable. 

CRC32C is defined as exactly 4 bytes. But I think most of java versions returns a long, is to 1) have compatibility with java.util.zip.Checksum and 2) to avoid signedness issues with putting a 4-byte quantity in a signed int. I think we can stick to 4 bytes or 8 bytes, it doesn't really matter here.

merlimat commented:
> close/reopen the pull request to trigger travis ci build as well.

@sijie I think for Travis you can trigger the rebuild in the Travis page directly (and it will update the PR, unlike Jenkins)

eolivelli commented on bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookieWriteLedgersWithDifferentDigestsTest.java:
```diff
@@ -93,21 +94,22 @@ public void setUp() throws Exception {
     }
 
-    @Test
+    @Test()
```
nit: remove parentheses

dlg99 replied:
done.

sijie commented:
@merlimat what is your thought about making the crc32c value exchangeable between pulsar implementation and java 9 implementation?

sijie commented:
I think this change is almost ready to go. I can incorporate into my PR #858 and move pulsar-checksum module into bookkeeper and make the crc32c value exchangeable between different implementations.

sijie commented:
@merlimat any thoughts here?

sijie commented:
ping @merlimat ?

sijie commented:
Can any other committers take a look at this? If we agree with CRC32C and 4 bytes, let's move forward with this.

eolivelli commented:
+1  for crc32c and 4 bytes from me

sijie commented:
My comment is merge this pull request if we agreed on crc32c and have the other pull requests to import pulsar-checksum.

eolivelli commented:
Ok for merging this
+1

sijie commented:
@merlimat can you review this when you have time?

sijie commented:
@merlimat when you have time, can you review this again? If you are good with this, let move forward with this patch.
